sinon/lib/stub.py

```diff
@@ -17,118 +17,178 @@ class SinonStub(SinonSpy):
+    def throws(self, exception=Exception):
+        """
+        Customizes the stub function to raise an exception. If conditions like withArgs or onCall
+        were specified, then the return value will only be returned when the conditions are met.
+
+        Args: exception (by default=Exception, it could be any customized exception)
+        Return: a SinonStub object (able to be chained)
+        """
+        def exception_function(*args, **kwargs):
+            raise exception
+        super(SinonStub, self).wrap2stub(exception_function)
         return self
 
-    def returns(self, obj):
+class SinonStubCondition(SinonStub):
+    """
+    Allows a new SinonStub object to be created each time the end user specifies a new condition. This is necessary
+    to mimic the behaviour of Sinon.JS.
+
+    Author: Jonathan Benn
+    """
+
+    def __new__(cls, *args, **kwargs):
+        """
+        Override the __new__ provided by SinonBase, since we don't want to do any function wrapping
         """
-        Customizing return of stub function
+        return object.__new__(cls)
 
-        The final chained functions of returns/throws will be triggered
-        If there is some conditions, it will ONLY triggered when meeting conditions
+    def __init__(self, copy_of=None, cond_args=None, cond_kwargs=None, oncall=None):
+        """
+        Args:
+            copy_of: the original SinonStub object that spawned this one
+            cond_args: the args to which a subsequent call to returns/throws should apply
+            cond_kwargs: the kwargs to which a subsequent call to returns/throws should apply
+            oncall: the integer call number to which a subsequent call to returns/throws should apply
+        """
+        self._copy_of = copy_of
+        self._cond_args = cond_args
+        self._cond_kwargs = cond_kwargs
+        self._oncall = oncall
+
+    def _get_original(self):
+        """
+        Returns the original SinonStub object that wrapped the function under test
+        """
+        return self._copy_of._get_original()
+
+    def returns(self, obj):
+        """
+        Customizes the return values of the stub function. If conditions like withArgs or onCall
+        were specified, then the return value will only be returned when the conditions are met.
 
         Args: obj (anything)
-        Return: self (able to be chained)
-        """
-        def return_function(*args, **kwargs):
-            """
-            A stub function with customized return
-            """
-            _ = args, kwargs
-            return obj
-
-        if self.cond_args or self.cond_kwargs or self.oncall:
-            self.__append_condition(return_function)
-            super(SinonStub, self).wrap2stub(self.stubfunc, self.condition)
-        else:
-            super(SinonStub, self).wrap2stub(return_function)
+        Return: a SinonStub object (able to be chained)
+        """
+        original = self._get_original()
```

Comment:
just `original = self._copy_of` is enough

Reply:
`_get_original()` is a recursive function. Let's say that the user makes a long call chain, such as `stub.withArgs(42).onFirstCall().returns(3)`, this call creates 2 new `SinonStubCondition` objects. If we break this down over multiple lines, it looks like this:

```python
stub = sinon.stub()
stubCondition1 = stub.withArgs(42)
stubCondition2 = stubCondition1.onFirstCall()
stubCondition2 = stubCondition2.returns(3)
```

If I am calling `stubCondition2._get_original()`, then it calls `stubCondition1._get_original()`, which calls `stub._get_original()`and then the `stub` returns `self`, and the `self` is propagated up the chain back up to `stubCondition2`.

With longer call chains even more new objects are created...

Maybe I should document this in `SinonStubCondition`?

Reply:
Agree! It's nice to have an explanation about the `return of function` and `description of API` in [Stub document](https://github.com/note35/sinon/blob/master/docs/Stubs.rst).

Reply:
I think `SinonStubCondition` should be renamed to `_SinonStubCondition`, because it should never be instantiated directly by the end user. It's just an internal implementation detail. Please see the [Pull Request](https://github.com/note35/sinon/pull/34) I created for this.